lib/widgets/compose_box.dart

```diff
@@ -269,19 +269,99 @@ class ComposeContentController extends ComposeController<ContentValidationError>
   }
 }
 
-class _ContentInput extends StatelessWidget {
+class _ContentInput extends StatefulWidget {
   const _ContentInput({
     required this.narrow,
+    required this.destination,
     required this.controller,
     required this.focusNode,
     required this.hintText,
   });
 
   final Narrow narrow;
+  final SendableNarrow destination;
   final ComposeContentController controller;
   final FocusNode focusNode;
   final String hintText;
 
+  @override
+  State<_ContentInput> createState() => _ContentInputState();
+}
+
+class _ContentInputState extends State<_ContentInput> with WidgetsBindingObserver {
+  String? _prevText;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.controller.addListener(_contentChanged);
+    widget.focusNode.addListener(_focusChanged);
+    // This listens to [AppLifecycleState] changes for us to check
+    // if the app is visible and in focus.
+    WidgetsBinding.instance.addObserver(this);
+  }
+
+  @override
+  void didChangeAppLifecycleState(AppLifecycleState state) {
+    super.didChangeAppLifecycleState(state);
+    switch (state) {
+      case AppLifecycleState.detached:
+      case AppLifecycleState.inactive:
+      case AppLifecycleState.hidden:
+      case AppLifecycleState.paused:
+        final store = PerAccountStoreWidget.of(context);
+        store.typingNotifier.stoppedComposing();
+      case AppLifecycleState.resumed:
+        // The app becoming visible and getting input focus doesn't
+        // necessarily mean that the user started typing, so do nothing.
+    }
+  }
+
+  @override
+  void didUpdateWidget(covariant _ContentInput oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
+      oldWidget.controller.removeListener(_contentChanged);
+      oldWidget.focusNode.removeListener(_focusChanged);
+      widget.controller.addListener(_contentChanged);
+      widget.focusNode.addListener(_focusChanged);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.controller.removeListener(_contentChanged);
+    widget.focusNode.removeListener(_focusChanged);
+    WidgetsBinding.instance.removeObserver(this);
+    super.dispose();
+  }
+
+  void _contentChanged() {
+    if (_prevText == widget.controller.text) {
+      // [TextEditingController] also notifies the listeners
+      // on selection updates.  Return early because we do not
+      // consider that as typing activity.
+      return;
+    }
+    final store = PerAccountStoreWidget.of(context);
+    (widget.controller.text.isEmpty)
+      ? store.typingNotifier.stoppedComposing()
+      : store.typingNotifier.keystroke(widget.destination);
```

Comment:
Rereading the API docs:
https://zulip.com/api/set-typing-status

the spec for what a client should do includes:
> While the user continues to actively type or otherwise interact with the compose UI (e.g. interacting with the compose box emoji picker), send regular `"op": "start"` requests to this endpoint

Thinking about that, here's a couple more situations where we should be sending such requests:
* The user hits one of the compose buttons. (Eventually if they do choose an image, etc., then that will cause a typing notice anyway because it'll edit the text in the input; but that may be many seconds later. In a fast-moving conversation, it may make a difference to let other participants know immediately that they're preparing something to send.)
* The text input's selection changes, as opposed to its text. That's what it'll look like to us if the user is moving their cursor around the input, preparing their next edit; or if they're selecting a part of their draft to cut and perhaps move around.
* Ideally, also if the user scrolls around the results of an autocomplete. If this one is at all complicated to implement, though, it's low priority — I think generally people are not likely to do this for more than a couple of seconds, between typing more letters to narrow down the autocomplete and then either picking a result or giving up.

Of those, the second one can be squashed into the main commit here, since it's just removing some logic that that one adds; the other two should probably be separate commits, just like your existing followup commit for app lifecycle changes.